### PR TITLE
DTLS v1.3: fix epoch 0 check on plaintext message

### LINKED ISSUE #6368: [Bug]: DTLS 1.3 start handshake with arbitrary epoch number
### Contact Details

conrad@owatz.de

### Version

5.6.0

### Description

In DTLS 1.3 the epoch number determines in which connection state you are an which key material to use.
Epoch 0 is used for unencrypzted messages, e.g. ClientHello or ServerHello. Epoch 2 is used for traffic protected by the handshake secret.
However, wolfSSL does not seem to check the epoch number of unencrypted traffic in DTLS 1.3. In my test I could send an unencrypted ClientHello or ServerHello with epoch number set to 2 and it got parsed normally. The handshake can succeed as long as the rest of the epoch numbers are sent correctly, meaning continuing with 2 for encrypzted handshake traffic.
The RFC states, that `The epoch number is initially zero and is incremented each time keying material changes and a sender aims to rekey. `

grafik

### Reproduction steps

To use DTLS 1.3 you have to enable it in user_setting.h:
```
#define WOLFSSL_DTLS
#define WOLFSSL_DTLS13
```

To reproduce the handshake I used an unreleased version of TLS-Attacker and modified the epoch value of the ClientHello or ServerHello.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by rizlik**
# Description

Fix a wrong check that allows client hello with epoch != 0 to get parsed. 
Fixes #6368
# Testing

A new test is added in the PR
Also, to test DLTSv1.2 downgrade:
```
./configure --enable-dtls --enable-dtls13 --enable-secure-renegotiation 
./example/server/server -vd -u -M -r
./example/client/client -v3 -u -R -r
```

### CODE REVIEW

**Comment by dgarske** [comment]
Retest this please

File: tests/api.c [hunk 1]
```diff
@@ -65649,6 +65649,67 @@ static int test_override_alt_cert_chain(void)
...
+    const int EPOCH_OFF = 3;
+    int ret, err;
+
+    XMEMSET(&test_ctx, 0, sizeof(test_ctx));
+    ret = test_memio_setup(&test_ctx, &ctx_c, &ctx_s, &ssl_c, &ssl_s,
+        wolfDTLSv1_3_client_method, wolfDTLSv1_3_server_method);
+    if (ret != 0)
+        return TEST_FAIL;
+
+    /* disable hrr cookie so we can later check msgsReceivece.got_client_hello
```
Comment (julek-wolfssl) [inline tests/api.c:65669]: `msgsReceivece` -> `msgsReceived`

File: tests/api.c [hunk 2]
```diff
@@ -65649,6 +65649,67 @@ static int test_override_alt_cert_chain(void)
...
+    test_ctx.s_buff[EPOCH_OFF + 1 ] = 0x2;
+
+    ret = wolfSSL_accept(ssl_s);
+    err = wolfSSL_get_error(ssl_s, ret);
+    if (ret == WOLFSSL_SUCCESS || err != WOLFSSL_ERROR_WANT_READ)
+        return TEST_FAIL;
+
+    if (ssl_s->msgsReceived.got_client_hello == 1)
+        return TEST_FAIL;
+
```
Comment (julek-wolfssl) [inline tests/api.c:65698]: Add a call to `test_memio_do_handshake` to make sure that we can recover from this.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
Nice job finding the fix. I really like the testing you added.

**Comment by rizlik** [comment]
Jenkins retest this please

**Review by julek-wolfssl — APPROVED** [review]

**Comment by JacobBarthelmeh** [comment]
Retest this please Jenkins. I noticed it has only 22 successful checks listed which is less than usual.

**Pull request merged**